### PR TITLE
support RFC 36 submission directives in `flux mini batch`

### PR DESCRIPTION
**Pull request by grondo**
This PR introduces support for processing [RFC 36](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_36.html) submission directives in the script argument to `flux mini batch`. This allows `flux mini batch` command line arguments to be embedded in the submission script. Options given on the command line at the time of submission will override embedded optoins.

This is a WIP while I write documentation, but figured I'd ask for an early review of the implementation in case there are needed changes before I document it.

### CODE REVIEW

**Comment by grondo** [comment]
Removing WIP since this is probably ready for some more serious review. I took a quick stab at documenting submission directives in `flux-mini(1)`, but still needs a proofread.

@ryanday36: We definitely want to get your approval of this feature before merging.

Note: you can experiment with the parsing by running the module standalone, in which the parser will just emit the directives it gleaned from whatever input file you give it, e.g. (in a builddir):

```console
$ cat batch.sh 
#!/bin/sh
#flux: -N4
#flux: --job-name=directive-test # set a job name
#flux: --setattr=user.foo='''
#flux: This is a
#flux: multiline string.
#flux: Sincerely,
#flux: 'flux mini batch'
#flux:'''
flux mini run -N4 hostname
$ src/cmd/flux python -m flux.job.directives batch.sh
SETARGS(['-N4'])
SETARGS(['--job-name=directive-test'])
SETARGS(["--setattr=user.foo=This is a\nmultiline string.\nSincerely,\n'flux mini batch'\n"])
```

File: src/bindings/python/flux/job/directives.py
Comment (grondo) [reply src/bindings/python/flux/job/directives.py:63]: I'll add a comment. A multiline is a string literal, so does not get any further processing by `shlex`.

File: src/cmd/flux-mini.py
Comment (grondo) [reply src/cmd/flux-mini.py:1816]: Yeah, I don't even remember adding this :thinking: :laughing: . I'll just drop it.

File: src/cmd/flux-mini.py
Comment (grondo) [reply src/cmd/flux-mini.py:1804]: Hm, I think the diff makes this a little confusing. The change is not independent, but is required because the function now processes a batch script which may contain directives, but if the `--wrap` option was given ,then the script cannot have directives since it was provided on the command line. So we just return. This check could be moved out a function into `main()`, but it seemed better to abstract the "script processing" into a single function.

Does that make sense?

Comment (chu11) [reply src/cmd/flux-mini.py:1804]: Ahh ok makes sense

File: src/bindings/python/flux/job/directives.py
Comment (grondo) [reply src/bindings/python/flux/job/directives.py:63]: Though now that I think about it, this may not be necessary. Multiline literals are passed to `shlex.quote()` before coming here, so I think the newlines should already be escaped. Let me test this.

Comment (grondo) [reply src/bindings/python/flux/job/directives.py:63]: Actually, nevermind. Multilines are appended as literals at the moment, so are not quoted (only triple quotes on a single line are quoted). It will be clearer and much less error prone to add a `disable_lexer` parameter or something to the `Directive` constructor that what is currently done, so let me make that change.

File: t/batch/directives/invalid/expected/001-invalid-directive.pattern
Comment (grondo) [reply t/batch/directives/invalid/expected/001-invalid-directive.pattern:1]: sheesh :facepalm:

**Comment by grondo** [comment]
Thanks @chu11! Must be some kind of typo record.

I think I've addressed all the comments thus far. As noted above, multi line quotes are now handled a bit better and more uniformly, and there is a better comment for the `Multiline` class.

File: src/bindings/python/flux/job/directives.py
Comment (chu11) [inline src/bindings/python/flux/job/directives.py:76]: maching -> matching

**Review by chu11 — APPROVED** [review]
LGTM, just one more typo!

File: src/bindings/python/flux/job/directives.py
Comment (grondo) [reply src/bindings/python/flux/job/directives.py]: Haha, ouch!

**Comment by grondo** [comment]
Thanks @chu11! I fixed that last typo.

Now I think we'd better wait until @ryanday36 gives his approval, then we can merge.

**Comment by ryanday36** [comment]
This all looks great to me @grondo! I didn't even find any typos.

**Comment by grondo** [comment]
Great, thanks for checking it out @ryanday36! I'll set MWP.

File: doc/man1/flux-mini.rst
Comment (garlick) [inline doc/man1/flux-mini.rst:861]: Might be good to add "in directives" to the comment character bullet (so people don't think they can prefix the sentinel with a comment)

File: doc/man1/flux-mini.rst
Comment (garlick) [inline doc/man1/flux-mini.rst:879]: I'd be fine if batch directives were never presented in upper case in our documentation.  Just saying :grin:

Comment (garlick) [reply doc/man1/flux-mini.rst:879]: Unless prefixed with a snarky comment like the "maximum nostalgia mode" one in the RFC.

File: doc/man1/flux-mini.rst
Comment (grondo) [reply doc/man1/flux-mini.rst]: Great suggestion!

**Comment by grondo** [comment]
I've addressed @garlick's comments and will now set MWP.

**Pull request merged**
